build/source/engine/allocspace.f90

```diff
@@ -106,36 +109,74 @@ subroutine allocate_gru_struc(nGRU,nHRU,err,message)
  mode=nf90_noWrite
  call nc_file_open(trim(infile), mode, ncid, err, cmessage)
  if(err/=0)then; message=trim(message)//trim(cmessage); return; endif
-
- ! get gru_ix dimension length
- err = nf90_inq_dimid(ncid, "ngru", gruDimID);             if(err/=0)then; message=trim(message)//'problem finding nGRU dimension'; return; endif
- err = nf90_inquire_dimension(ncid, gruDimID, len = nGRU); if(err/=0)then; message=trim(message)//'problem reading nGRU dimension'; return; endif
-
- ! get hru_dim dimension length (ie., the number of hrus in entire domain)
- err = nf90_inq_dimid(ncid, "nhru", hruDimID);             if(err/=0)then; message=trim(message)//'problem finding nHRU dimension'; return; endif
- err = nf90_inquire_dimension(ncid, hruDimID, len = nHRU); if(err/=0)then; message=trim(message)//'problem reading nHRU dimension'; return; endif
  
- allocate(gru_struc(nGRU), index_map(nHRU), stat=err)
- if(err/=0)then; err=20; message=trim(message)//'problem allocating space for mapping structures'; return; endif
+ if (strtHRU>0) then
+  !this is a subset run. assuming the user know the dimemsions
+  
+ else
+  
+  ! get hru_dim dimension length (ie., the number of hrus in entire domain)
+  err = nf90_inq_dimid(ncid, "nhru", hruDimID);             if(err/=0)then; message=trim(message)//'problem finding nHRU dimension'; return; endif
+  err = nf90_inquire_dimension(ncid, hruDimID, len = nHRU); if(err/=0)then; message=trim(message)//'problem reading nHRU dimension'; return; endif
+ 
+ end if 
  
- allocate(gru_id(nGRU),hru_id(nHRU),hru2gru_id(nHRU),stat=err)
+ ! read the hruID and HRU's gruID
+ allocate(gru_id(nHRU),hru_id(nHRU),hru2gru_id(nHRU),index_map(nHRU),stat=err)
  if(err/=0)then; err=20; message=trim(message)//'problem allocating space for zLocalAttributes gru-hru correspondence vectors'; return; endif
-
- ! read gru_id from netcdf file
- err = nf90_inq_varid(ncid, "gruId", varid);     if(err/=0)then; message=trim(message)//'problem finding gruId variable'; return; endif
- err = nf90_get_var(ncid,varid,gru_id);          if(err/=0)then; message=trim(message)//'problem reading gruId variable'; return; endif
-
- ! read the HRU to GRU mapping
+ 
+ err = nf90_inq_varid(ncid, "hruId", varid); if(err/=0)then; message=trim(message)//'problem finding hruId variable'; return; endif
+ err = nf90_get_var(ncid,varid,hru_id,start=(/max(strtHRU,1)/),count=(/nHRU/))
+ if(err/=0)then; message=trim(message)//'problem reading hruId variable'; return; endif  
+ 
  err = nf90_inq_varid(ncid, "hru2gruId", varid); if(err/=0)then; message=trim(message)//'problem finding hru2gruId variable'; return; endif
- err = nf90_get_var(ncid,varid,hru2gru_id);      if(err/=0)then; message=trim(message)//'problem reading hru2gruId variable'; return; endif
+ err = nf90_get_var(ncid,varid,hru2gru_id,start=(/max(strtHRU,1)/),count=(/nHRU/))
+ if(err/=0)then; message=trim(message)//'problem reading hru2gruId variable'; return; endif
+ 
+ ! extract unique gruID 
+ nGRU = 1
```

Comment:
I think it would be good to add more comments in this section, similar to the removed code in the read_params subroutine.  That way people can easily figure out what is going on to fill the GRU-HRU mapping structures.